### DISCUSSION TITLE
How to pass custom void elements to rehype-parse?

### DISCUSSION THREAD

**Discussion by AshtarCodes** [rehypejs/rehype#74] [Q&A] [answered]
I see that rehype-stringify has an options object with a "voids" property, but I can't find anything similar for rehype-parse. 

I'm parsing 100s of html documents that contain web components, and some of these are written as voids (without a closing tag). When I inspect the parsed tree, I can see that a sibling component will be inserted into the node.children property of the previous component. This results in the first sibling wrapping the second sibling component. This does not happen if elements have opening and closing tags. 

 Here is the source file: 
![Source HTML file](https://user-images.githubusercontent.com/74083145/140374506-0484f64d-3dce-4e4b-a66d-87d3c37d0523.png)

Here is the resulting output. Note that the mds-button is wrapping the mds-alert.
![Transformed HTML file](https://user-images.githubusercontent.com/74083145/140374705-42a6d431-6042-4204-996a-935dd85916c7.png)

This is how I've configured the processor. "parse" is rehype-parse, and "toHtml" is rehype-stringify. "transformComp" is a custom plugin. 
![Unified Processor config](https://user-images.githubusercontent.com/74083145/140374977-601cb9f1-a315-48ad-9e43-348e624b686e.png)

Any advice or resource you can point me to will be greatly appreciated.

**Comment by wooorm** [comment, marked as answer]
No, there’s no such option. It’s not valid HTML if the non-voids are not closed. `rehype-parse` is an HTML parser.
I recommend using correct HTML, and thus closing your custom elements, so that both browsers and rehype can handle them.

P.S. your input shows `blueButton` and the like, which are also not valid custom HTML elements (they’d need to include a `-` to become one)

**Reply by wooorm** [reply]
are you dealing with Vue or Angular or so?

**Reply by AshtarCodes** [reply]
The framework we are converting from is a custom one that uses ractive.js underneath it. That's probably why the syntax is a little wonky. Currently, there are 100s or even 1000+ files written in this way. The new framework uses lit element or lit html i think. The point of my tool is to assist with a rewrite of all these files. Do you have any recommendations on how to rewrite these files so that rehype can parse them?  

I'm wondering if I use rehype-wrap to isolate each component in a div before parsing (and then removing it in the tree), that could be a workaround. Or is there anyway I can extend the rehype parser to support custom void elements?

**Reply by wooorm** [reply]
Hasn't been updated in a few years but `@starptech/rehype-webparser` might be able to parse it: https://github.com/Prettyhtml/prettyhtml/tree/HEAD/packages/rehype webparser.
Then for the rest you can use "normal" rehype to format it with closing tags and such

**Reply by AshtarCodes** [reply]
Okay, thank you @wooorm. I'll look into this. I appreciate the quick response.
